Add the ability to match bound claims using globs

randomswdev opened the pull request:
This merge request is intended to add the ability to match bound claims not only using an exact string matching, but also a mathcing based on regular expressions.
Added a new flag `bound_claims_are_regexps`; if it is set to true the vaule of each entry in the `bound_claims` map is expected to be a string (not an array of strings) that will be interpreted as a regular expresison. The syntax for the regular expressions is the golang syntax.

This would allow mathcing JWT whose claims, for example, contain randomly generated portions of text.

Solves issue https://github.com/hashicorp/vault-plugin-auth-jwt/issues/82

kalafut commented:
Hi. Thanks for the contribution. We have generally avoided regexes in Vault, especially in the areas dealing with authentication or authorization. The main reason is that regexes are notoriously easy to get wrong, which is this case could result in unauthorized access.

Globs, on the other hand, _are_ used in Vault and are a more user-friendly way to handle many of the same use cases. Allowing globs as bound claims would be a useful improvement if you want to revise the PR.

Some specific feedback:
* Retaining an opt-in parameter is good, though I think something like `bound_claim_format` would be more flexible. Allowed today would be `"string"`, `""` (i.e. unspecified, treated as atring), or `"blob"`. This would allow future expansion, e.g. to regexes if we found a compelling reason to use them
* This blob library is recommended as it's used elsewhere in Vault: https://github.com/ryanuber/go-glob
* Is there any reason to restrict the expected claim to a single pattern, as opposed to potentially a list of patterns as is supported today?

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/vault-plugin-auth-jwt?pullRequest=89)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/vault-plugin-auth-jwt?pullRequest=89)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/vault-plugin-auth-jwt?pullRequest=89).</sub>

randomswdev commented:
> Hi. Thanks for the contribution. We have generally avoided regexes in Vault, especially in the areas dealing with authentication or authorization. The main reason is that regexes are notoriously easy to get wrong, which is this case could result in unauthorized access.
> 
> Globs, on the other hand, _are_ used in Vault and are a more user-friendly way to handle many of the same use cases. Allowing globs as bound claims would be a useful improvement if you want to revise the PR.
> 
> Some specific feedback:
> 
>     * Retaining an opt-in parameter is good, though I think something like `bound_claim_format` would be more flexible. Allowed today would be `"string"`, `""` (i.e. unspecified, treated as atring), or `"blob"`. This would allow future expansion, e.g. to regexes if we found a compelling reason to use them
> 
>     * This blob library is recommended as it's used elsewhere in Vault: https://github.com/ryanuber/go-glob
> 
>     * Is there any reason to restrict the expected claim to a single pattern, as opposed to potentially a list of patterns as is supported today?

I modified my pull request to address all of your comments:

*  Instead of a boolean, I added a type string. Accepted values are `string` and `glob`
*  I used the glob validation library that you suggested
*  If the claim type is `glob` it is still possible associating to a claim name a list of glob expressions

Thanks for the feedback and let me know if you think additional changes are needed.

kalafut commented on path_role.go:
```diff
@@ -105,6 +108,11 @@ Defaults to 60 (1 minute) if set to 0 and can be disabled if set to -1.`,
 			},
+			"bound_claims_type": {
+				Type:        framework.TypeString,
+				Description: `How to interpret values in the map of claims/values (which must match for login): allowed values are string or glob`,
```
```suggestion
				Description: `How to interpret values in the map of claims/values (which must match for login): allowed values are 'string' or 'glob'`,
```

randomswdev replied:
Implemented the change in a new commit (to simply reviewing it)

kalafut commented on claims.go:
```diff
@@ -89,7 +90,13 @@ func validateAudience(boundAudiences, audClaim []string, strict bool) error {
 // met in allClaims.
-func validateBoundClaims(logger log.Logger, boundClaims, allClaims map[string]interface{}) error {
+func validateBoundClaims(logger log.Logger, boundClaimsType string, boundClaims, allClaims map[string]interface{}) error {
+	if boundClaimsType != boundClaimsTypeString && boundClaimsType != boundClaimsTypeGlob {
```
Since we're checking for a valid type when the role is created, I think checking again here in an internal function is probably unnecessary.

randomswdev replied:
Removed the check in the new commit (to simply reviewing it)

kalafut commented on path_role.go:
```diff
@@ -426,8 +436,31 @@ func (b *jwtAuthBackend) pathRoleCreateUpdate(ctx context.Context, req *logical.
 	}
 
+	boundClaimsType := data.Get("bound_claims_type").(string)
+	if boundClaimsType != boundClaimsTypeString && boundClaimsType != boundClaimsTypeGlob {
```
Suggested refactoring:

```
switch boundClaimsType {
    case boundClaimsTypeString, boundClaimsTypeGlob:
        role.BoundClaimsType = boundClaimsType
    default:
        return logical.ErrorResponse("invalid 'bound_claims_type': %s", boundClaimsType), nil
}
```

randomswdev replied:
Implemented the change in the new commit (to simply reviewing it)

kalafut commented on path_role.go:
```diff
@@ -426,8 +436,31 @@ func (b *jwtAuthBackend) pathRoleCreateUpdate(ctx context.Context, req *logical.
+				claimsValuesList, ok := normalizeList(claimValues)
+
+				if !ok {
+					return logical.ErrorResponse("received claim is not a string or list: %v", claimValues), nil
```
```suggestion
					return logical.ErrorResponse("claim is not a string or list: %v", claimValues), nil
```

randomswdev replied:
Removed the check in the new commit (to simply reviewing it)

kalafut commented on path_role.go:
```diff
@@ -426,8 +436,31 @@ func (b *jwtAuthBackend) pathRoleCreateUpdate(ctx context.Context, req *logical.
+
+				for _, claimValue := range claimsValuesList {
+					if _, ok := claimValue.(string); !ok {
+						return logical.ErrorResponse("received claim is not a string: %v", claimValue), nil
```
```suggestion
						return logical.ErrorResponse("claim is not a string: %v", claimValue), nil
```

randomswdev replied:
Removed the check in the new commit (to simply reviewing it)

kalafut commented:
Add upgrade login to the internal role read function (`b.role`), ala: 

```
if boundClaimsType == "" {
    boundClaimsType = boundClaimsTypeString
}
```

kalafut commented on claims_test.go:
```diff
@@ -382,7 +400,8 @@ func TestValidateBoundClaims(t *testing.T) {
+			boundClaimsType: "string",
 			boundClaims: map[string]interface{}{
 				"email": "d",
 			},
```
Include a test case with `boundClaimsType == ""` since that's what all existing roles will be read as.

randomswdev replied:
Extedned an existing test case by removing the `bound_claims_type` entry from the json representation of a role, saving the role in storage and reading it back (same testing paatern as for the role_type attribute).
I completely missed the "upgrade" path, i.e. reading from storage an existing role. Thanks for catching it.

kalafut commented:
@randomswdev The PR is looking really good. Just a few small comments, plus the need to handle existing roles.

randomswdev commented:
> 
> 
> Add upgrade logic to the internal role read function (`b.role`), ala:
> 
> ```
> if boundClaimsType == "" {
>     boundClaimsType = boundClaimsTypeString
> }
> ```

Added the suggested logic in the new commit (to simply reviewing it)

randomswdev commented:
I should have addressed all the review comments. I commited the changes in a separate commit to simplify further reviewing them.
Let me know if additional changes are needed.

kalafut commented on path_role_test.go:
```diff
@@ -670,6 +673,26 @@ func TestPath_Read(t *testing.T) {
+	}
+
+	// Run read test for "upgrade" case. The legacy role is not changed in storage, but
+	// reads will populate the `bound_claims_type` with "jwt".
```
```suggestion
	// reads will populate the `bound_claims_type` with "string".
```

randomswdev replied:
Addressed in the third commit, sorry for the mistake.

kalafut commented on claims.go:
```diff
@@ -112,10 +115,25 @@ func validateBoundClaims(logger log.Logger, boundClaims, allClaims map[string]in
+			if useGlobs {
+				vs, ok := v.(string)
+				if !ok {
+					return fmt.Errorf("received claim is not a glob string: %v", v)
```
```suggestion
					return fmt.Errorf("received claim is not a string: %v", v)
```

kalafut commented on claims.go:
```diff
@@ -112,10 +115,25 @@ func validateBoundClaims(logger log.Logger, boundClaims, allClaims map[string]in
-					break scan
+			if useGlobs {
+				vs, ok := v.(string)
+				if !ok {
```
We've already verified this at role update/create time, right?  I think `vs := v.(string)` is appropriate. We'd panic if not string, but that's OK since the role code shouldn't never let non-strings become expected values.

randomswdev replied:
Addressed in the third commit.

randomswdev commented on path_role.go:
```diff
@@ -426,8 +440,33 @@ func (b *jwtAuthBackend) pathRoleCreateUpdate(ctx context.Context, req *logical.
+	case boundClaimsTypeString, boundClaimsTypeGlob:
+		role.BoundClaimsType = boundClaimsType
+	default:
+		return logical.ErrorResponse("invalid 'bound_claims_type': %s", boundClaimsType), nil
```
Added three new test cases in the 4th commit.

randomswdev commented on claims_test.go:
```diff
@@ -391,9 +410,97 @@ func TestValidateBoundClaims(t *testing.T) {
+			},
+			errExpected: true,
+		},
+		{
```
Added the test case in the 4th commit.

kalafut commented:
@randomswdev  Thanks!
